### ISSUE TITLE
router problems while copying content of dist folder to a subdir of a static server

### ISSUE THREAD

**Issue by orange-log** [coryhouse/react-slingshot#458]
Node version: v7.2.0

npm version: 3.10.9

Operating system: Win7 64

Command line used: cmd

Steps to reproduce: clone repo, change publicPath in webpack.config.prod.js to "./", run "npm run build", copy content of dist folder to a subdir on a static server, observer that the routing is broken.

**Comment by kwelch** [comment]
`publicPath` should be an absolute path. If you are deploying to a location that is in a sub directory, then the `publicPath` should match the path of the app with run on a webserver.

**Comment by orange-log** [comment]
I changed the `publicPath` path to `"/path/to/"` and I am calling it like `localhost/path/to/`. Still I become redirect to the 404 page of the app. Also I tried to configure the path in `routes.js` `<Route path="/path/to/" component={App}>`. With this configuration the homepage of the app is shown correctly but a call on `localhost/path/to/asdasd` is not showing the 404 page as I would expect. The described behavior is working fine with the default configuration `/` and `npm start` and `npm run build` in the root directory.

Thx for helping!

**Comment by coryhouse** [comment]
If you want to support deep linking, you need to set up wildcard routing on your webserver. (redirect all traffic to index.html so that client-side routing takes over).

**Comment by orange-log** [comment]
unfortunately I can't do such settings on the webserver. Does it mean I can't run the app in a subdir? A colleague of mine suggested to switch from browser history to hash history. would that be a reasonable option? thx.

**Comment by kwelch** [comment]
I don't believe HashRouter would address the deep routing. What server are you using?

**Comment by orange-log** [comment]
atm an ordinary apache server on my localhost to simulate the production server

**Comment by coryhouse** [comment]
If you're running Apache, you should be able to place a .httpd.conf file in the root directory to configure wildcard routing, regardless of host.

**Comment by coryhouse** [comment]
Correction, declare a .htacess file in the relevant dir

**Comment by orange-log** [comment]
yes that would work. I'll consider to give it a shot. thx a lot!

**Comment by kwelch** [comment]
Awesome. We have an open item to add documentation around deployments. We would love a write up on how you were able to get this to work. 🎉

**Comment by nickytonline** [comment]
For reference, the open issue for deployment documentation is #408.

**Comment by orange-log** [comment]
its not very nice, but its working now:

- in `webpack.config.prod.js`: `publicPath: '/path/to/',`
- in `routes.js`:

`
[...]
import { useRouterHistory } from 'react-router'
import { createHistory } from 'history'

const history = useRouterHistory(createHistory)({
  basename: '/path/to/'
})

export default (
  <Route history={history} path="/path/to/" component={App}>
    [...]
  </Route>
);
`

- any Link `<Link to="/path/to/target">link</Link>`
- in `/path/to/.htaccess`:

`
<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteBase /
    RewriteRule ^index\.html$ - [L]
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteRule . /path/to/index.html [L]
</IfModule>

`

**Comment by lwilli** [comment]
@orange-log Thanks for the instructions for deploying to a static site. I'm trying to deploy my project to GitHub Pages and it's proven to be a challenge. I'm still having trouble with my routes though, neither the About Page nor the NotFoundPage ever render.

    <Route history={history} path="/ProjectName/" component={App}>
        <IndexRoute component={MainPage}/>
        <Route path="about" component={AboutPage}/>
        <Route path="*" component={NotFoundPage}/>
    </Route>

App and MainPage render fine, but I can't get to AboutPage (which I believe should be at `lwilli.github.io/ProjectName/about`) and NotFoundPage (which I believe should be any nonexistent page such as `lwilli.github.io/ProjectName/nothinghere`). 

If I'm deploying to `lwilli.github.io/ProjectName/`, how should I configure my routes so they actually work? (Also, I can't add the .htaccess file since it's on GH Pages.) Thanks in advance!

**Comment by kwelch** [comment]
With GH pages, I typically treat the 404 page as my index. So in a case like this I would rename `index.html` to `404.html` then when routing tries to find a page it will fail and display the custom 404 which is where you app will be running. 

I hope that helps.
